### DISCUSSION TITLE
Debugging a CMake dependency

### DISCUSSION THREAD

**Discussion by simgt** [mesonbuild/meson#11228] [Q&A] [answered]
I'm trying to link [an example program](https://pytorch.org/tutorials/advanced/cpp_export.html#step-3-loading-your-script-module-in-c) against libtorch with Meson, using [Torch's provided CMake modules](https://github.com/pytorch/pytorch/blob/master/cmake/TorchConfig.cmake.in). I'm on macOS. The original CMake-based example builds and runs fine:

```cmake
# Configure with: cmake -DCMAKE_PREFIX_PATH=(python3 -c "import torch; print(torch.utils.cmake_prefix_path)") -B build
cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
project(libtorch-example)
find_package(Torch REQUIRED)
add_executable(example-app-cmake example-app.cpp)
message("${TORCH_LIBRARIES}")
target_link_libraries(example-app-cmake "${TORCH_LIBRARIES}")
set_property(TARGET example-app-cmake PROPERTY CXX_STANDARD 14)
```

I try to build the same code with Meson in the following way:

```python
# Configure with: meson build --cmake-prefix-path=(python3 -c "import torch; print(torch.utils.cmake_prefix_path)")
project('libtorch-example', ['c', 'cpp'], default_options : ['cpp_std=c++14'])
torch = dependency('Torch', version : '>=1.10', method : 'cmake')
message(torch.get_variable(cmake : 'TORCH_LIBRARIES'))
executable('example-app-meson', 'example-app.cpp', dependencies : torch)
```

The program produced by meson fails to load libc10 that Torch provides, and I guess libtorch_cpu would follow.

The difference of linking in-between the two programs:

```
> otool -L build/example-app-meson
build/example-app-meson:
@rpath/libc10.dylib (compatibility version 0.0.0, current version 0.0.0)
@rpath/libtorch_cpu.dylib (compatibility version 0.0.0, current version 0.0.0)
/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.36.0)
/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)

> otool -L build/example-app-cmake
build/example-app-cmake:
/opt/homebrew/opt/libtorch/lib/libc10.dylib (compatibility version 0.0.0, current version 0.0.0)
/opt/homebrew/opt/libtorch/lib/libtorch.dylib (compatibility version 0.0.0, current version 0.0.0)
/opt/homebrew/opt/libtorch/lib/libtorch_cpu.dylib (compatibility version 0.0.0, current version 0.0.0)
/opt/homebrew/opt/protobuf/lib/libprotobuf.32.dylib (compatibility version 33.0.0, current version 33.4.0)
/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.36.0)
/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
```

In meson, `TORCH_LIBRARIES` prints `torch torch_library`
whereas with cmake I get `torch;torch_library;/opt/homebrew/lib/libc10.dylib;/opt/homebrew/lib/libkineto.a`

The build seems to find c10 however:
```
> cat build/meson-private/cmake_Torch/CMakeCache.txt | grep c10
c10_LIBRARY:FILEPATH=/opt/homebrew/lib/libc10.dylib
```

**Would you have any suggestions on how I could understand what's not working with the meson build?** I'd like to recover some messages from the underlying `TorchConfig.cmake` but didn't find any obvious way to make the configuration verbose enough.

**Comment by simgt** [comment, marked as answer]
My issue was specifically linked to https://github.com/mesonbuild/meson/issues/3882, not sure what the motive is behind this feature.
Regarding debugging a dependency, it seems that there is barely anything outside of a few prints in `meson-logs/` and `meson-private/`: https://github.com/mesonbuild/meson/issues/4563

**Reply by lterrac** [reply]
Ehy @simgt I am having the exact same issue! What solution did work for you specifically? :)
